// Copyright 2023 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package xls;

import "google/protobuf/duration.proto";
import "xls/ir/transform_metrics.proto";

message PassResultProto {
  // How many times the pass was run.
  optional int64 run_count = 1;
  // How many runs changed the IR.
  optional int64 changed_count = 2;
  // Aggregate transformation metrics across the runs.
  // TODO(allight): Consider holding de-aggregated metrics?
  optional TransformMetricsProto metrics = 3;
  // Total duration of the running of the pass.
  optional google.protobuf.Duration pass_duration = 4;
}

// Overall metrics for a pass pipeline.
message PipelineMetricsProto {
  // Map from pass short_name to overal metrics for that pass.
  map<string, PassResultProto> pass_results = 1;
}
